controllers/submission.go

```diff
@@ -50,28 +34,114 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
 		return http.StatusMethodNotAllowed, nil
 	}
 
+	var body = struct {
+		Code     string
+		Language string
+	}{}
+
+	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
+		return http.StatusBadRequest, err
+	}
+
 	p, ok := passenger.FromContext(ctx)
 	if !ok {
 		return http.StatusUnauthorized, nil
 	}
 
 	resultKey, err := datastore.DecodeKey(mux.Vars(r)["resultKey"])
+	if err != nil {
+		return http.StatusNotFound, err
+	}
 
 	if !util.HasParent(p.User, resultKey) {
 		return http.StatusBadRequest, errors.New("cannot submit answer for other users")
 	}
 
 	taskKey, err := datastore.DecodeKey(mux.Vars(r)["taskKey"])
-	// Note: When more task kinds are added, see controllers.CreateFinalResult.
-	switch taskKey.Kind() {
-	case model.CodeTaskKind:
-		return runner.HandleCodeSubmission(ctx, w, r, resultKey, taskKey)
-	// TODO(victorbalan, flowlo): Use correct kind when possible.
-	case "QuestionTask":
-		return http.StatusInternalServerError, errors.New("question submissions are not yet implemented")
-	default:
-		return http.StatusBadRequest, errors.New("Unknown submission kind.")
+	if err != nil {
+		return http.StatusNotFound, err
+	}
+
+	var task model.Task
+	if err = datastore.Get(ctx, taskKey, &task); err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	// Furthermore, the name of the GCS object is derived from the of the
+	// encapsulating Submission. To avoid race conditions, allocate an ID.
+	low, _, err := datastore.AllocateIDs(ctx, model.SubmissionKind, resultKey, 1)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	submissionKey := datastore.NewKey(ctx, model.SubmissionKind, "", low, resultKey)
```

Comment:
`key.Encode()` will generate an opaque string representation of the key, so (a) we cannot rely on it, as it might change to some different format and (b) it does not generate a nice tree-like structure AFAIK. It has some slashes in there, but won't be as nice as `nameObject`.